torch/_functorch/_aot_autograd/functional_utils.py

```diff
@@ -191,7 +191,79 @@ def has_metadata_mutation(f_arg, arg, *, check_only_storage_mutation: bool):
         return has_metadata_mutation_
 
 
-def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires_grad):
+def gen_alias_from_base(
+    aliased_base_tensor,
+    target_meta_tensor,
+    target_requires_grad,
+    target_functional_tensor=None,
+):
+    # Patch the correct requires_grad field of the output tensor, depending on whether:
+    # (i) the reconstructed output (out) was came from a tensor that requires grad or not;
+    # and (ii) the concrete returned output does require grad or not.
+    def patch_requires_grad(out):
+        if aliased_base_tensor.requires_grad and not target_requires_grad:
+            out = out.detach()
+        elif not aliased_base_tensor.requires_grad and target_requires_grad:
+            out.requires_grad_(True)
+        return out
+
+    # If provided, use the target functional tensor for replaying the views.
+    #
+    # In summary, we use the fact that FunctionalTensorWrapper saves the view
+    # functions applied to itself (collected during functionalization) so as
+    # to replay them (view functions) on the aliased_base_tensor.
+    if target_functional_tensor is not None:
+        functional_tensor = target_functional_tensor.tensor
+
+        # First, retrieve the actual base tensor.
+        base = torch._functionalize_base(functional_tensor)
+
+        # Then, apply the necessary ViewMeta so that we reach the offset.
+        # These are ViewMeta that should already be applied to aliased_base_tensor.
+        if target_functional_tensor.offset > 0:
+            base = torch._functionalize_apply_view_metas(
+                target_functional_tensor.tensor,
+                base=base,
+                start=0,
+                end=target_functional_tensor.offset,
+            )
+
+        if base.shape == aliased_base_tensor.shape:
+            # Only apply these functions if the new base (aliased_base_tensor) has the
+            # same shape as the old base. The idea is:
+            #
+            #   1. target_functional_tensor represents the functional tensor that corresponds
+            #      to target_meta_tensor output
+            #
+            #   2. In order to make the new base have the same shape as target_functional_tensor,
+            #      the base of the latter (old base) should have the same shape as the new base
+            #      (that's what FunctionalTensorWrapper guarantees with its regenerate_from_base)
+            #
+            #   3. Otherwise, we have no guarantees the output will have the same shape as
```

Comment:
hmm.. it's not clear to me from the description when this does / does not happen (when `base.shape != aliased_base_tensor.shape`. Do you have a description?

Reply:
This is a sanity check. I left an `if` here, because: (i) the other cases (not using functional tensor) do the same check; (ii) in the case we get it wrong with `ViewMeta`, we can just fallback to some other thing.

Reply:
Explicitly not supporting in-place view operations, so marking as resolved.